Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -107,178 +129,195 @@ static void EFB_Write(u32 data, u32 addr)
 	}
 }
 
+u32 dbat_table[1 << (32 - 17)];
+u32 ibat_table[1 << (32 - 17)];
 
 static void GenerateDSIException(u32 _EffectiveAddress, bool _bWrite);
 
-template <XCheckTLBFlag flag, typename T>
-__forceinline T ReadFromHardware(const u32 em_address)
+template <XCheckTLBFlag flag, typename T, bool never_translate = false>
+__forceinline static T ReadFromHardware(u32 em_address)
 {
-	int segment = em_address >> 28;
-	// Quick check for an address that can't meet any of the following conditions,
-	// to speed up the MMU path.
-	if (!BitSet32(0xCFC)[segment])
+	if (!never_translate && UReg_MSR(MSR).DR)
 	{
-		// TODO: Figure out the fastest order of tests for both read and write (they are probably different).
-		if ((em_address & 0xC8000000) == 0xC8000000)
+		auto translated_addr = TranslateAddress<flag>(em_address);
+		if (!translated_addr.valid)
 		{
-			if (em_address < 0xcc000000)
-				return EFB_Read(em_address);
-			else
-				return (T)mmio_mapping->Read<typename std::make_unsigned<T>::type>(em_address);
-		}
-		else if (segment == 0x8 || segment == 0xC || segment == 0x0)
-		{
-			return bswap((*(const T*)&m_pRAM[em_address & RAM_MASK]));
-		}
-		else if (m_pEXRAM && (segment == 0x9 || segment == 0xD || segment == 0x1))
-		{
-			return bswap((*(const T*)&m_pEXRAM[em_address & EXRAM_MASK]));
+			if (flag == FLAG_READ)
+				GenerateDSIException(em_address, false);
+			return 0;
 		}
-		else if (segment == 0xE && (em_address < (0xE0000000 + L1_CACHE_SIZE)))
+		if ((em_address & (HW_PAGE_SIZE - 1)) > HW_PAGE_SIZE - sizeof(T))
 		{
-			return bswap((*(const T*)&m_pL1Cache[em_address & L1_CACHE_MASK]));
+			// This could be unaligned down to the byte level... hopefully this is rare, so doing it this
+			// way isn't too terrible.
+			// TODO: floats on non-word-aligned boundaries should technically cause alignment exceptions.
+			// Note that "word" means 32-bit, so paired singles or doubles might still be 32-bit aligned!
+			u32 em_address_next_page = (em_address + sizeof(T) - 1) & ~(HW_PAGE_SIZE - 1);
+			auto tlb_addr_next_page = TranslateAddress<flag>(em_address_next_page);
+			if (!tlb_addr_next_page.valid)
+			{
+				if (flag == FLAG_READ)
+					GenerateDSIException(em_address_next_page, false);
+				return 0;
+			}
+			T var = 0;
+			u32 tlb_addr = translated_addr.address;
+			for (u32 addr = em_address; addr < em_address + sizeof(T); addr++, tlb_addr++)
+			{
+				if (addr == em_address_next_page)
+					tlb_addr = tlb_addr_next_page.address;
+				var = (var << 8) | ReadFromHardware<flag, u8, true>(tlb_addr);
+			}
+			return var;
 		}
+		em_address = translated_addr.address;
 	}
 
-	if (bFakeVMEM && (segment == 0x7 || segment == 0x4))
+	// TODO: Make sure these are safe for unaligned addresses.
+
+	// Locked L1 technically doesn't have a fixed address, but games all use 0xE0000000.
+	if ((em_address >> 28) == 0xE && (em_address < (0xE0000000 + Memory::L1_CACHE_SIZE)))
```

Comment:
I guess it would be possible... but it's kind of complicated.  It would require much more code to implement of dcbz_l, we couldn't do fastmem for load/store to locked L1, we'd have to do an extra hashtable lookup for every non-fastmem load, and we'd have to have special handling in the JIT and the fastmem arena for locked addresses which overlap other locations we care about.  None of it is hard, exactly, but it adds a lot of complexity, and we'd probably lose a bit of performance.

Not sure how much of that to put into a comment in the code; most of it follows naturally from the fact that we're messing with the behavior of load/store from arbitrary addresses.